kt_jvm_grpc.bzl

```diff
@@ -0,0 +1,145 @@
+    args.add_all(direct_descriptor_set)
+    args.add("--")
+    args.add_all(transitive_descriptor_set)
+
+    ctx.actions.run(
+        outputs = [output_dir],
+        inputs = depset(transitive = [direct_descriptor_set, transitive_descriptor_set]),
+        arguments = [args],
+        executable = ctx.executable._generator,
+        mnemonic = "KtGrpcGenerator",
+        progress_message = "Generating Kotlin gRPC extensions for %s" % proto_dep.label,
+    )
+
+def _build_srcjar(ctx, proto_dep, input_dir, source_jar):
+    args = ctx.actions.args()
+    args.add("c")
+    args.add(source_jar.path)
+    args.add_all(depset([input_dir]))
+    ctx.actions.run_shell(
+        outputs = [source_jar],
+        inputs = [input_dir],
+        tools = [ctx.executable._zipper],
+        arguments = [args],
+        command = "{zipper} $1 $2 $3".format(zipper = ctx.executable._zipper.path),
+        mnemonic = "KtGrpcSrcJar",
+        progress_message = "Generating Kotlin gRPC srcjar for %s" % proto_dep.label,
+    )
+
+def _kt_grpc_generate_code_impl(ctx):
+    proto_dep = ctx.attr.srcs[0]
+    name = ctx.label.name
+
+    gen_src_dir_name = "%s/ktgrpc" % name
+    gen_src_dir = ctx.actions.declare_directory(gen_src_dir_name)
+    source_jar = ctx.actions.declare_file("%s.srcjar" % name)
+
+    _invoke_generator(ctx, proto_dep, gen_src_dir)
+    _build_srcjar(ctx, proto_dep, gen_src_dir, source_jar)
+
+    java_info = ctx.attr.deps[0][JavaInfo]
+    default_info = DefaultInfo(files = depset([source_jar, gen_src_dir]))
+
+    return [java_info, default_info]
+
+_kt_grpc_generate_code = rule(
+    attrs = dict(
+        srcs = attr.label_list(
+            providers = [ProtoInfo],
+        ),
+        deps = attr.label_list(
+            providers = [JavaInfo],
+        ),
+        _zipper = attr.label(
+            executable = True,
+            cfg = "host",
+            default = Label("@bazel_tools//tools/zip:zipper"),
+            allow_files = True,
+        ),
+        _generator = attr.label(
+            default = Label("//compiler/src/main/java/io/grpc/kotlin/generator:GeneratorRunner"),
+            cfg = "host",
+            executable = True,
+        ),
+    ),
+    fragments = ["java"],
+    provides = [JavaInfo],
+    implementation = _kt_grpc_generate_code_impl,
+)
+
+def kt_jvm_grpc_library(
+        name,
+        srcs = None,
+        deps = None,
+        tags = None,
+        testonly = None,  # None to not override Blaze's default for //javatests, b/112708042
+        compatible_with = None,
+        restricted_to = None,
+        visibility = None,
+        flavor = None,
+        deprecation = None,
+        features = []):
+    """This rule compiles Kotlin APIs for gRPC services from the proto_library in srcs.
+
+    Args:
+      name: A name for the target
+      srcs: Exactly one proto_library target to generate Kotlin APIs for
+      deps: Exactly one java_grpc_library target for srcs[0]
```

Comment:
Could we make the dep on the java_proto_library, rather than the java_grpc_library?

The dependency on the java_grpc_library is mostly an implementation detail that might change in future, while the java_proto_library dep is an API choice we want to commit to.

The `@StubFor` annotation might make breaking this dependency hard in the future, but I'd like to do what we can to retain that option.

Reply:
@lowasser are you proposing that this build macro creates an internal `java_grpc_library` for now and then that will be removed if you break the dependency? (An earlier revision did this, actually.)

Reply:
Yes, that's the request.

Reply:
Done.